web-ui/badges.ml

```diff
@@ -0,0 +1,80 @@
+type schema = {
+  schema_version : int; [@key "schemaVersion"]  (** always: 1 *)
+  label : string;
+  message : string;
+  color : string option;  (** default: "lightgrey" *)
+  label_color : string option; [@key "labelColor"]  (** default: "grey" *)
+  is_error : bool option; [@key "isError"]  (** default: false *)
+  named_logo : string option; [@key "namedLogo"]  (** default: none *)
+  logo_svg : string option; [@key "logoSvg"]  (** default: none *)
+  logo_color : string option; [@key "logoColor"]  (** default: none *)
+  logo_width : string option; [@key "logoWidth"]  (** default: none *)
+  logo_position : string option; [@key "logoPosition"]  (** default: none *)
+  style : string option;  (** default: "flat" *)
+  cache_seconds : int option;  (** default: 300 *)
+}
+[@@deriving yojson]
+(** Extracted from https://shields.io/endpoint. *)
+
+let v ~label ~message ?color ?label_color ?is_error ?named_logo ?logo_svg
+    ?logo_color ?logo_width ?logo_position ?style ?cache_seconds () =
+  {
+    schema_version = 1;
+    label;
+    message;
+    color;
+    label_color;
+    is_error;
+    named_logo;
+    logo_svg;
+    logo_color;
+    logo_width;
+    logo_position;
+    style;
+    cache_seconds;
+  }
+
+let schema_of_status =
+  let v = v ~label:"ocaml-ci" in
+  function
+  | `Not_started -> v ~message:"unknown" ~color:"inactive" ()
+  | `Pending -> v ~message:"building" ~color:"yellow" ()
+  | `Failed -> v ~message:"failing" ~color:"critical" ()
+  | `Passed -> v ~message:"passing" ~color:"success" ()
+
+let ( let* ) = Lwt.Infix.( >>= )
+
+open Lwt.Infix
+module Capability = Capnp_rpc_lwt.Capability
+module Client = Ocaml_ci_api.Client
+module Server = Cohttp_lwt_unix.Server
+
+let normal_response = Lwt.map (fun x -> `Response x)
+
+let respond_error status body =
+  let headers = Cohttp.Header.init_with "Content-Type" "text/plain" in
+  Server.respond_error ~status ~headers ~body () |> normal_response
+
+let ( let*! ) x f =
```

Comment:
Would be useful to log the detailed error for the ocaml-ci admin (and maybe just return a generic error to the badge system).

Reply:
Done in 46fb09c.